app/controllers/concerns/foreman/controller/authentication.rb

```diff
@@ -17,15 +17,13 @@ def authenticate
         logger.info("Authorized user #{user.login}(#{user.to_label})")
         set_current_user user
       else
-        if api_request?
-          false
-        else
-          # Keep the old request uri that we can redirect later on
-          session[:original_uri] = request.fullpath
-          @available_sso ||= SSO::Base.new(self)
-
-          redirect_to @available_sso.login_url unless @available_sso.has_rendered
-        end
+        return false if api_request?
+        # Keep the old request uri that we can redirect later on
+        session[:original_uri] = request.fullpath
+        @available_sso ||= SSO::Base.new(self)
+        return if @available_sso.has_rendered
+        flash[:warning] = _('Your session has expired.') if session[:user] && !User.current
```

Comment:
1. Please also add an assertion to the test for the flash message (`assert_equal "...", flash[:warning]`)
1. Should `session[:user]` be deleted at this point? It's now known that the user ID is incorrect, it might be safest to wipe the entire contents - even better, using `backup_session_content { reset_session }` in the same way as `expire_session` (app/controllers/concerns/foreman/controller/session.rb).
1. The message should probably match `expire_session`'s message by adding ", please login again".